### ISSUE TITLE
Using lookahead to detect generic JSX elements?

### ISSUE THREAD

**Issue by isiahmeadows** [Microsoft/TypeScript#6395]
_Initially from a [couple](https://github.com/Microsoft/TypeScript/issues/3960#issuecomment-165375803) [different](https://github.com/Microsoft/TypeScript/issues/3960#issuecomment-132192858) comments, but elaborated here._

Currently, in order to use a generic JSX element, you must alias both the type and interface to a non-generic specialization.

``` ts
class Select<T> extends React.Component<SelectProps<T>, any> {}

// Alias
type StringSelect = new () => Select<string>
const StringSelect = <StringSelect> Select

const Form = () => <StringSelect items={['a','b']} />
```

It would be a _lot_ more convenient and consistent to use something like this:

``` ts
const Form = () => <Select<string> items={['a', 'b']} />
```

In order to differentiate the latter, you can use a single token of lookahead in the parser to know if it's a generic JSX element:
- If the next token is a `<`, like in `< Select <`, then consume the rest of the type as a generic type. After consuming that:
  - If the next token is a `>`, then finish the type as a generic type cast.
  - If the next token is an identifier or a `{` (for spread attributes, when they get implemented), then parse the rest as a JSX element.
  - Otherwise, parse the rest as a type cast.
- If the next token is a `>`, like in `< Select >`, then finish the type as a simple type cast.
- If the next token is an identifier or a `{`, then parse the rest as a JSX element.
- Otherwise, consume the rest as a type cast.

I don't know of any reason that wouldn't be feasible to do, considering async arrow functions, which are [already implemented](https://github.com/Microsoft/TypeScript/pull/5231), require potentially a whole call expression of parsing to differentiate between `let f = async (x, y)`, which is a call expression, and `let f = async (x, y) => x`, which is an async arrow function. (I feel sorry for whoever ends up implementing that in V8, which doesn't support any backtracking or much lookahead for memory reasons.)

/cc @RyanCavanaugh

**Comment by RyanCavanaugh** [comment]
We've gotten this request a number of times and it does seem reasonable in terms of parsing.

**Comment by RyanCavanaugh** [comment]
Accepting PRs.

This should be easy to parse but I think it's going to be some rather difficult work in the checker.

**Comment by stepancar** [comment]
@RyanCavanaugh whats status of this issue?
It's very impotant functionality.

```javascript

import * as React from 'react';

interface MyTestProps<T> {
    values: Array<T>;
    selectHandler: (selectedVal: T) => void;
}

export class MyTest<T> extends React.Component<MyTestProps<T>, {}> {

    render() {
        return (
            <ul>
                { this.props.values.map(val =>
                    <li>{val}</li>
                ) }
            </ul>
        );
    }
}

class TestMyTest extends React.Component<{}, {}> {
    render() {
        return (
            <MyTest values={[1, 2, 3, 4]} selectHandler={(val) => { } } />
        );
    }
}

```

compiler throws error "type number[] is not assignable to type T[]"

But in this example

```javascript

import * as React from 'react';

interface MyTestProps<T> {
    values: Array<T>;
    selectHandler: (selectedVal: T) => void;
}

export class MyTest<T> extends React.Component<MyTestProps<T>, {}> {

    render() {
        return (
            <ul>
                { this.props.values.map(val =>
                    <li>{val}</li>
                ) }
            </ul>
        );
    }
}

new MyTest({ values: ['a', 'b'], selectHandler: (v) => { v.length } });

```
selectHandler param v have type string;

**Comment by Liero** [comment]
meanwhile, I use this shortcut to alias the non generic specialization:

Alias:
`const StringSelect = Select as new () => Select<String>;`

Usage:
`const Form = () => <StringSelect items={['a', 'b']} />`

**Comment by donaldpipowitch** [comment]
@RyanCavanaugh Could this be part of a future planned milestone? 2.1 or 2.2?

**Comment by RyanCavanaugh** [comment]
We'll be basically redoing JSX typechecking in 2.1 or 2.2 and can probably look at this at that time.

**Comment by donaldpipowitch** [comment]
Cool. Thank you for the update! :)

**Comment by bang88** [comment]
Any update ?

**Comment by alitaheri** [comment]
@RyanCavanaugh Please consider supporting at least type inference when possible. No need to add new syntax. Just having a generic component infer its type parameters solve a huge number use-cases.

**Comment by RyanCavanaugh** [comment]
@alitaheri I believe the existing JSX work going on at #12107 will enable that scenario

**Comment by MastroLindus** [comment]
Still not working on typescript@2.2.1.
The typescript team is making an incredibly good job making typescript and react such a good combination, but this issue is still quite painful.
The workaround works fine (
const BooleanSelect: new () => SingleColumnSelect<boolean | null> = SingleColumnSelect;
<BooleanSelect values={[null, true, false]}/>
) but it's not only ugly and un-elegant, it is also really unclear for newcomers to react&typescript

Either solution (allowing to specify the generic inside jsx or using type inference) would be really really welcome

**Comment by zzzev** [comment]
@MastroLindus in the meantime you might consider this workaround, which I find to be a bit simpler to read and write:

```
class Select<T> extends React.Component<...> {...}
class BooleanSelect extends Select<boolean> {}
class SomeComponent extends React.Component<{}, {}> {
  render() {
    return <BooleanSelect />; // this works fine
  }
}
```

**Comment by MastroLindus** [comment]
@zzzev I am aware of it( or others, like just a simple cast) but to define a new class just for the workaround somehow it feels even dirtier to me, even if in practice the two forms are not that much different... I simply cannot find one that I am really satisfied about

**Comment by yuit** [comment]
We are still need a way to infer type parameter in the original example for generic React class component but here are some possible work around in some cases:

(Note this is for 2.3 compiler)
** Using SFC **
as mentioned by @RyanCavanaugh's [comment](https://github.com/Microsoft/TypeScript/issues/6395#issuecomment-273544921), compiler now can infer type parameter in SFC
```ts
import * as React from "react";

type SelectProps<T > = { items: T[] }
function Select<T>(attr: SelectProps<T>) {...}

const someItems = [1, 2, 3, 4];
const someItems2 = ["ss"];

function renderJsx() {
    return <Select items={someItems} />; // Ok
}
```

**Nightly & future 2.3.3**
Using default type parameter (in this example, it may not be desirable to do so but the compiler will not give you error for being generic)
```ts
import * as React from "react";

type SelectProps<T > = { items: T[] }
class Select<T = any> extends React.Component<SelectProps<T>, any> { }

const someItems = [1, 2, 3, 4];
const someItems2 = ["ss"];

function renderJsx() {
    return <Select items={someItems} />; // OK
}
```

**Comment by MichaelTontchev** [comment]
Would love to have this feature - would enable a lot of good scenarios :)

**Comment by ianks** [comment]
Is this likely to ever be merged?

**Comment by MichaelTontchev** [comment]
Would also like to ping this topic.

I'm implementing a smoother way of handling forms in react and it really depends on having generic types. The workaround brings back much of the boilerplate I'd like to avoid :/

**Comment by goloveychuk** [comment]
what I've noticed:
```ts
import * as React from 'react';

interface Props<T> {
  a: T
  b: T[]
}

function Comp1<T>(props: Props<T>) {
  return null
}


class Comp2<T> extends React.Component<Props<T>> {
  render(){
    return null
  }
}

const props = {
  a: '',
  b: [2],
}

const a = <Comp1 {...props} /> //error here

React.createElement(Comp1, props)

const b = <Comp2 {...props} />

React.createElement(Comp1, props)
```
So, createElement every time passing T=any.
When using jsx and component is Stateless Component - checking works. So, workaround for now could be to wrap your generic components with Stateless comp.

**Comment by AlexGalays** [comment]
Still as annoying as ever :D

**Comment by FredyC** [comment]
So it's more than a year now since the issue was opened, we are going for a Typescript 2.8 and is not even on [the roadmap](https://github.com/Microsoft/TypeScript/wiki/Roadmap)? I don't get it...

**Comment by antanas-arvasevicius** [comment]
maybe not so trivial to do

**Comment by isiahmeadows** [comment]
@antanas-arvasevicius Apparently, [the checker is the difficult part](https://github.com/Microsoft/TypeScript/issues/6395#issuecomment-178263804). If you read my initial bug, it was me addressing one implementation aspect mentioned in comments there.

**Comment by RyanCavanaugh** [comment]
This should in theory become easier (possibly even trivial) once @weswigham 's work to refactor JSX typechecking is complete.

I checked, and idle complaining that we haven't implemented popular a feature request has not caused new engineer time to appear.

**Comment by weswigham** [comment]
We actually do inference and generic instantiation for generic JSX components as of our last release, so we don't even actually really _need_ anything more than the parse step for this anymore (I mean, ofc the call resolution need to be modified to check for type arguments before doing inference, but still - more done than before).

**Comment by weswigham** [comment]
Also, the OP is a tad out of date; Unless I'm mistaken, we don't even support parsing angle bracket style typecasts in JSX files, so the described ambiguity wouldn't exist.

**Comment by RyanCavanaugh** [comment]
`<T>e` expressions aren't supported in JSX. The only exception to `<T>` in a leading expression position starting a JSX element is a generic arrow function `<T>(x: T) => e`, which I believe has some limited lookahead to try to disambiguate (or maybe we required you to write `T extends {}` ? I'd have to check).

**Comment by weswigham** [comment]
We require a constraint on the first type parameter.

**Comment by RyanCavanaugh** [comment]
I stand corrected, idle complaining actually did improve the situation 😆

**Comment by antanas-arvasevicius** [comment]
:D how about other issues related to TSX? Especially deprecating magic in
global JSX namespace and instead using type inference directly from
createElement? Allowing multiple JSX definitions would allow to use not
only react, but create many html like DSL's for e.g. database schema
definitions, permission roles, state machines which "config like
structures" could be directly translated to js.

**Comment by weswigham** [comment]
@antanas-arvasevicius We've merged code that allows us to look up the current global types in [a namespace local to the factory function instead of globally](https://github.com/Microsoft/TypeScript/pull/22207) already, which should be a significant improvement once libraries move to utilizing it.

We _want_ to move to just using the factory function's signature, but there's a slight problem - to create a perfectly well-typed factory function that we can actually do inferences for, you need `ReturnType` (and `InstanceType` and `FirstArgument`) conditional type-types that are overload-aware (otherwise you need compiler internal magic that assumes how the factory function works or looks things up in a namespace, like we do now). As is, we could basically do it, but we'd break overloading on stateless function components (we'd always pick the last signature), which would be no bueno; so fully moving away from the `JSX` namespace (anywhere) might require that we either improve conditional types or implement real "call" types first. We're actively looking at it, though!

**Comment by antanas-arvasevicius** [comment]
Thank you for review of what's going on, it is more complicated/interesting
than it looked :) One year ago I had a trouble with JSX that I couldn't
declare that ElementType would be dependable on InstanceType e.g. <List>
would return real instance of object which is list and not an ADT like
structure as react vdom because not all frameworks works in vdom mode
extjs, isomorphic smartclient etc, but their GUI could be represented in
jsx way, my opened issue
https://github.com/Microsoft/TypeScript/issues/13890
Please consider it also, and if factory function is difficult/impossible to
implement maybe that proposal for ElementType could be released?

**Comment by sebmarkbage** [comment]
Since this is not compatible with the JSX spec as it is right now, it would’ve been good to have the discussion over on the JSX spec repo.

JSX already allows for nested elements in the attribute position `<Foo foo=<Foo /> />` so it’s possible this wouldn’t play well with other additions. I don’t see an immediate problem but would be good to take a pass over there in the future like we agreed to do for other additions.

**Comment by weswigham** [comment]
> I don’t see an immediate problem but would be good to take a pass over there in the future like we agreed to do for other additions.

Apologies - we've had this issue open for over two years with a pretty straightforward design desired from the beginning (implementation of typechecking being the real blocker, which we quietly implemented without the syntax in 2.7); I wasn't aware we had (or needed, since it's a type-based syntax addition, like type arguments for call or new expressions) any renewed outreach set up around this (probably something to do with me not having worked on it 3 years ago XD).

> JSX already allows for nested elements in the attribute position `<Foo foo=<Foo /> />` 

As an aside, occasionally I've heard people familiar with the jsx spec point that out (within the team, I think it's a favorite "I bet you didn't know x was valid jsx" fact). We still don't parse it - it's been years since we first noticed it and _still_ haven't gotten a serious issue for it. [Still seems underused.](https://github.com/facebook/jsx/issues/53) It _is_ [still broken in babel 6.](https://babeljs.io/repl/#?babili=false&browsers=&build=&builtIns=false&code_lz=MYewdgzgLgBAHjAvDAPAMRCGAzTj2YwD0AfMSUA&debug=false&forceAllTransforms=false&shippedProposals=false&circleciRepo=&evaluate=true&fileSize=false&lineWrap=false&presets=es2015%2Creact%2Cstage-2&prettier=false&targets=&version=6.26.0&envVersion=) (fixed in the 7 beta though, so it's got that going for it, though [fragments are broken right now](http://babeljs.io/repl/build/master#?babili=false&browsers=&build=&builtIns=false&code_lz=MYewdgzgLgBAHjAvDAPAMRCGAzTiUB8KAJgJYBuA9EdTNUA&debug=false&forceAllTransforms=false&shippedProposals=false&circleciRepo=&evaluate=false&fileSize=false&lineWrap=true&presets=react&prettier=false&targets=&version=7.0.0-beta.46&envVersion=)) Given how it was [spoken of the last time it came up](https://github.com/facebook/jsx/pull/94), its future isn't exactly golden (FYI, babylon7 is the only parser on ASTExplorer that can parse the fragment production, but babel can't actually transpile it right now. heh.).

But it was brought up while I implemented this, IIRC 😉 We don't think there are any obvious conflicts (beyond the ones we already have to disambiguate with leading generic parameters and occasionally left shift token that occur elsewhere), either (unless someone's going to start suggesting silly things like jsx elements as property names in fragments or something). If you're looking for a brief on how it was implemented, in our case it's just, after the `JSXElementName` production in opening/self closing tags,  (optionally) a one-token lookahead for `<`, followed by our `TypeArgumentList` production (I'm not sure how you'd try to represent leaving space for such a thing in the JSX spec) and a required closing `>` (before continuing with standard JSX production).
